File: frontend/types/social.ts
// ─────────────────────────────────────────────
//  HyperFitness Phase 4 — Social & Feed Types
// ─────────────────────────────────────────────

export interface UserProfile {
  id: string;
  name: string;
  avatarUrl: string;
  username: string;
  isVerified?: boolean;
  level?: number;
  rank?: string;
  followersCount: number;
  followingCount: number;
}

export type PostMediaType = "image" | "video" | "workout_share" | "achievement";

export interface SocialPost {
  id: string;
  author: UserProfile;
  content: string;
  mediaUrl?: string;
  mediaType?: PostMediaType;
  likesCount: number;
  commentsCount: number;
  isLikedByMe: boolean;
  isSavedByMe: boolean;
  createdAt: string;
  workoutData?: {
    exercise: string;
    reps: number;
    duration: string;
    formScore: number;
  };
}

export interface Comment {
  id: string;
  author: UserProfile;
  content: string;
  likesCount: number;
  createdAt: string;
}

export interface SocialNotification {
  id: string;
  type: "like" | "comment" | "follow" | "challenge_invite" | "achievement";
  actor: UserProfile;
  message: string;
  isRead: boolean;
  createdAt: string;
  link?: string;
}

export interface TrainerProfile extends UserProfile {
  bio: string;
  specialties: string[];
  rating: number;
  clientCount: number;
  availableSlots: number;
  plans: TrainerPlan[];
}

export interface TrainerPlan {
  id: string;
  title: string;
  description: string;
  price: number;
  durationWeeks: number;
}

File: backend/app/models/social.py
# ─────────────────────────────────────────────
#  HyperFitness Phase 4 — Social Models
# ─────────────────────────────────────────────
from sqlalchemy import Column, String, DateTime, Integer, ForeignKey, Text, Boolean, JSON, Index
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
import uuid
from datetime import datetime

from app.database.session import Base
from app.models.models import TimestampMixin

class Post(Base, TimestampMixin):
    __tablename__ = "posts"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    content = Column(Text, nullable=False)
    media_url = Column(String(255), nullable=True) # Cloudinary URL
    media_type = Column(String(50), nullable=True) # image, video, workout_share
    workout_id = Column(UUID(as_uuid=True), ForeignKey("workouts.id", ondelete="SET NULL"), nullable=True)
    likes_count = Column(Integer, default=0)
    comments_count = Column(Integer, default=0)
    
    # Indexes for feed queries
    __table_args__ = (
        Index("ix_posts_created_at_user", "created_at", "user_id"),
    )

class Comment(Base, TimestampMixin):
    __tablename__ = "comments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    post_id = Column(UUID(as_uuid=True), ForeignKey("posts.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    content = Column(Text, nullable=False)

class Like(Base, TimestampMixin):
    __tablename__ = "likes"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    post_id = Column(UUID(as_uuid=True), ForeignKey("posts.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    
    __table_args__ = (
        Index("ix_likes_post_user", "post_id", "user_id", unique=True),
    )

class Follow(Base, TimestampMixin):
    __tablename__ = "follows"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    follower_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    following_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    __table_args__ = (
        Index("ix_follows_follower_following", "follower_id", "following_id", unique=True),
    )

class Community(Base, TimestampMixin):
    __tablename__ = "communities"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name = Column(String(255), nullable=False)
    description = Column(Text, nullable=True)
    banner_url = Column(String(255), nullable=True)
    member_count = Column(Integer, default=0)
    created_by = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="SET NULL"), nullable=True)

class TrainerProfile(Base, TimestampMixin):
    __tablename__ = "trainer_profiles"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, unique=True)
    bio = Column(Text, nullable=True)
    specialties = Column(JSON, nullable=True)
    rating = Column(Integer, default=5)
    client_count = Column(Integer, default=0)
    is_verified = Column(Boolean, default=False)
